### PR TITLE
Migrate tcl into aura -> 80badger-20wbtc

### PR DESCRIPTION
**Pull request by petrovska-petro**
Tackles #595 

```
brownie run issue/595/migrate_tcl
```

**Note:** decide to leverage the inheritance of the convex class as it is similar SC architecture. Added some area for methods which are not relevant. Modified a bit balancer for the unstake method, at later time we can continue with the refurbishment of either class wanted to keep it practical and minimised changed within one pr.

### CODE REVIEW

File: scripts/issue/595/migrate_tcl.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    (_, _, _, rewards) = vault.aura.get_pool_info(bpt)
+    rewards = vault.contract(rewards)
+
+    vault.balancer.unstake_all(bpt)
+
+    bpt_amount = bpt.balanceOf(vault)
+
+    bal_before_stake = rewards.balanceOf(vault)
+    vault.aura.deposit_and_stake(bpt, bpt_amount)
```
Comment (jorijnsmit) [inline scripts/issue/595/migrate_tcl.py:23]:
```suggestion
    bal_before_stake = rewards.balanceOf(vault)
    vault.aura.deposit_all_and_stake(bpt)
```
prevents dusty issues by internally leveraging `aura.booster.depositAll`

File: great_ape_safe/ape_api/balancer.py [hunk 2]
```diff
@@ -280,6 +280,20 @@ def stake_all(self, pool, destination=None, dusty=False):
         self.stake(pool, mantissa, destination, dusty)
 
 
+    def unstake(self, pool, mantissa, claim=True):
+        gauge = self.safe.contract(self.gauge_factory.getPoolGauge(pool))
+        bal_pool_before = pool.balanceOf(self.safe)
```
Comment (jorijnsmit) [inline great_ape_safe/ape_api/balancer.py:285]: yes good catch here on checking pool balance instead of gauge balance 👍

**Review by jorijnsmit — COMMENTED** [review]
runs successfully but left some minor comments

File: great_ape_safe/ape_api/aura.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class Aura(Convex):
+    def __init__(self, safe):
+        self.safe = safe
+
+        self.aura = interface.ERC20(r.treasury_tokens.AURA, owner=self.safe.account)
+        self.aura_bal = interface.ERC20(
+            r.treasury_tokens.AURABAL, owner=self.safe.account
+        )
```
[the remark this answers is not in the document]
Comment (petrovska-petro) [reply great_ape_safe/ape_api/aura.py]: good suggestion, initially i thought on them more on the usual erc20 usual checkups of `transfer`, `balanceOf`...etc. but makes sense to grab all custom logic 👍

**Review by shuklaayush — APPROVED** [review]
LGTM

File: great_ape_safe/ape_api/balancer.py [hunk 4]
```diff
@@ -280,6 +280,20 @@ def stake_all(self, pool, destination=None, dusty=False):
...
+        gauge.withdraw(mantissa, claim)
+        assert pool.balanceOf(self.safe) >= bal_pool_before
+
+
+    def unstake_all(self, pool, claim=True):
+        gauge = self.safe.contract(self.gauge_factory.getPoolGauge(pool))
+        bal_pool_before = pool.balanceOf(self.safe)
+        gauge.withdraw(gauge.balanceOf(self.safe), claim)
+        assert pool.balanceOf(self.safe) >= bal_pool_before
```
Comment (GalloDaSballo) [inline great_ape_safe/ape_api/balancer.py:294]: If gauges are 1-1 shouldn't this be exactly 
`gauge_bal = gauge.balanceOf(self.safe)`
`pool.balanceOf(self.safe) == bal_pool_before + gauge_bal )`
?

File: great_ape_safe/ape_api/balancer.py [hunk 5]
```diff
@@ -280,6 +280,20 @@ def stake_all(self, pool, destination=None, dusty=False):
         self.stake(pool, mantissa, destination, dusty)
 
 
+    def unstake(self, pool, mantissa, claim=True):
+        gauge = self.safe.contract(self.gauge_factory.getPoolGauge(pool))
+        bal_pool_before = pool.balanceOf(self.safe)
+        gauge.withdraw(mantissa, claim)
+        assert pool.balanceOf(self.safe) >= bal_pool_before
```
Comment (GalloDaSballo) [inline great_ape_safe/ape_api/balancer.py:287]: See comment on Unstake all, if 1-1 then we can have a stricter check

File: scripts/issue/595/migrate_tcl.py (hunk 1 above)
Comment (GalloDaSballo) [inline scripts/issue/595/migrate_tcl.py]: I can't seem to find the code for this function, can you please link?
Aura seems like it doesn't have this function:
https://github.com/Badger-Finance/badger-multisig/blob/2cb800d5babb226a1cde1d3a9129fcbd91864aa1/great_ape_safe/ape_api/aura.py

File: great_ape_safe/ape_api/balancer.py (hunk 5 above)
Comment (petrovska-petro) [reply great_ape_safe/ape_api/balancer.py:287]: sorted on #606 as well

**Pull request merged**